### PR TITLE
syscalls: Add system call for cache flush & invalidate

### PR DESCRIPTION
**Pull request by aasthagr**
include/cache.h: System calls declaration and implementation
kernel/cache_handlers.c: Defination of verification functions

Signed-off-by: Aastha Grover <aastha.grover@intel.com>

### CODE REVIEW

**Review by tyrantlizard77 — COMMENTED** [review]
The sys_cache_flush() and sys_cache_invd() functions are what need to be system calls.

But we also know that some arches do not require a system call to be made in order to run these operations. What do we have with respect to this on x86 and ARC?

**Comment by aasthagr** [comment]
> The sys_cache_flush() and sys_cache_invd() functions are what need to be system calls.
> 
> But we also know that some arches do not require a system call to be made in order to run these operations. What do we have with respect to this on x86 and ARC?

Changed the names of the system calls to sys_cache_flush() & sys_cache_invd(). And for the latter question, I think if an architecture doesnt require a system calls, then it should call the implementation function directly from my understanding.

File: include/syscall_handler.h [hunk 1]
```diff
@@ -539,7 +539,12 @@ static inline int z_obj_validation_check(struct z_object *ko,
 #define Z_SYSCALL_OBJ_NEVER_INIT(ptr, type) \
 	Z_SYSCALL_IS_OBJ(ptr, type, _OBJ_INIT_FALSE)
 
+__syscall void sys_cache_flush(void *addr, size_t size);
+
+__syscall void sys_cache_invd(void *addr, size_t size);
+
```
Comment (ceolin) [inline include/syscall_handler.h:545]: aren't these functions declared in `include/cache.h` ? It seems that your re-declaring them.

**Review by ceolin — CHANGES_REQUESTED** [review]
Hey @aasthagr please confirm whether or not what I've said is right. Also they should be declared in that header IMHO.

File: include/syscall_handler.h (hunk 1 above)
Comment (aasthagr) [reply include/syscall_handler.h:545]: Those are the API names and are declared in include/cache.h. These are the system calls. Will these not be considered different? Although the names are same.

File: kernel/thread.c [hunk 2]
```diff
@@ -922,5 +922,29 @@ static inline k_ticks_t z_vrfy_k_thread_timeout_expires_ticks(
 	Z_OOPS(Z_SYSCALL_OBJ(t, K_OBJ_THREAD));
 	return z_impl_k_thread_timeout_expires_ticks(t);
 }
+
+void z_impl_sys_cache_flush(void *addr, size_t size)
```
Comment (nashif) [inline kernel/thread.c:926]: why are those being added in kernel/thread.c? How is this related to threads?

File: include/syscall_handler.h (hunk 1 above)
Comment (nashif) [reply include/syscall_handler.h:545]: Make something a syscall needs to be done where the API is declared, in this case in include/cache.h AFAIK.

File: include/syscall_handler.h [hunk 3]
```diff
@@ -539,7 +539,12 @@ static inline int z_obj_validation_check(struct z_object *ko,
...
 #include <driver-validation.h>
+#include <syscalls/syscall_handler.h>
```
Comment (nashif) [inline include/syscall_handler.h:547]: why is this needed?

**Review by nashif — CHANGES_REQUESTED** [review]

File: include/syscall_handler.h (hunk 1 above)
Comment (aasthagr) [reply include/syscall_handler.h]: Addressed this.

File: kernel/thread.c (hunk 2 above)
Comment (aasthagr) [reply kernel/thread.c]: Was just a place to start with. Have shifted this code to a new file kernel/cache_handlers.c.

File: include/syscall_handler.h (hunk 3 above)
Comment (aasthagr) [reply include/syscall_handler.h]: Earlier I was doing this as system calls were declared in this file. Corrected this and the declaration of system calls is now in include/cache.h.

**Review by tyrantlizard77 — APPROVED** [review]

**Comment by andrewboie** [comment]
LGTM!

File: kernel/CMakeLists.txt [hunk 4]
```diff
@@ -58,6 +58,7 @@ target_sources_ifdef(
   mem_domain.c
   userspace_handler.c
   userspace.c
+  cache_handlers.c
```
Comment (nashif) [inline kernel/CMakeLists.txt:61]: I do not think this belongs under kernel/, this is not part of the kernel and should probably be part of the arch/?

**Review by nashif — CHANGES_REQUESTED** [review]

**Review by ceolin — APPROVED** [review]

File: kernel/CMakeLists.txt (hunk 4 above)
Comment (andrewboie) [reply kernel/CMakeLists.txt:61]: not following you here. to me arch/ is the architecture-specific parts of kernel/. if it's not arch-specific it doesn't go there. And we are exposing this functionality as a system call, which is a kernel entry point.

Comment (nashif) [reply kernel/CMakeLists.txt:61]: want to see the handlers where the code is, in this case it is in arch and declared in its own header, so it is not a "kernel" (in my view, is what is really implemented under kernel/) and is rather an os/architecture feature.
Any reason why this cant go into arch/common/?

Besides, based on the comment 4 lines above this change, the ordering is wrong if we are to keep it in this location.

Comment (andrewboie) [reply kernel/CMakeLists.txt:61]: > "kernel" (in my view, is what is really implemented under kernel/) and is rather an os/architecture feature.

I'm not fully understanding the difference in semantics from "os" and "kernel" here. This is implementing interfaces defined at the kernel level. If these weren't inline functions they would need the same C file.

I also don't understand why "arch/common" exists other than past bikeshedding. arch/ is supposed to be for the parts of the kernel which require different implementations based on CPU type

I really do not want to go down a path where we have some arbitrary distinction between "kernel/" and "arch/common/" that nobody can precisely explain. IMO everything in arch/common belongs in the kernel

Comment (nashif) [reply kernel/CMakeLists.txt:61]: just like we have drivers/i2c/i2c_handlers.c where the i2c syscalls are implemented and not under kernel/, for the sake of having things organized, put the handlers at the same level where the implementation is.

Comment (andrewboie) [reply kernel/CMakeLists.txt:61]: but the implementation is in `include/sys/`. I don't see what `arch/common/` has to do with this. The sycalls for include/sys/atomic.c are implemented in the kernel. Every syscall in `include/sys` is implemented in either `kernel` or `lib/os` as far as I can see.

You might argue, "put it in lib/os" but we shouldn't be putting syscalls there either, the ones that are in there need to be moved to the kernel. When we eventually support separate kernel/application, the lib/os stuff will all go in the application and not the kernel's memory space

Comment (aasthagr) [reply kernel/CMakeLists.txt]: > Besides, based on the comment 4 lines above this change, the ordering is wrong if we are to keep it in this location.

Addressed this. The ordering of cache_handlers.h is before the last two files now in kernel/CMakeLists.txt.

**Review by nashif — APPROVED** [review]
